### DISCUSSION TITLE
Any plans to offer rosbag-like API?

### DISCUSSION THREAD

**Discussion by chengguizi** [eclipse-ecal/ecal#818] [General]
I understand eCAL relies on HDF5 standard to save its measurement / frames into files.

I would like to know if there is a designed way to programmatically read / write the content of the recorded files? something like rosbag.write() / rosbag::View

http://wiki.ros.org/rosbag/Code%20API

My search gives me some clue like:

https://eclipse-ecal.github.io/ecal/_api/classeCAL_1_1eh5_1_1HDF5Meas.html#class-hdf5meas
https://github.com/eclipse-ecal/ecal/blob/67cf3ec9e0616f7abe6ee5014ec89b681e3ef414/app/rec/rec_client_core/src/job/hdf5_writer_thread.cpp#L183

But it does not look like a high level API to use yet. Out of curiousity, any suggestions on what's the best way to go about read / write .hdf5 in eCAL, similar to read / write .bag in ROS?

**Comment by rex-schilasky** [comment]
Just to complete your search results. These are the high level C++ HDF5/Protobuf API samples:

https://github.com/eclipse-ecal/ecal/blob/master/samples/cpp/measurement/measurement_write/src/measurement_write.cpp
https://github.com/eclipse-ecal/ecal/blob/master/samples/cpp/measurement/measurement_read/src/measurement_read.cppt

Here a python sample writing and reading hdf5 files using the basic measurement API:

https://github.com/eclipse-ecal/ecal/blob/master/samples/python/ecalhdf5_rw/ecalhdf5_rw.py

And finally a python sample reading HDF5 using a higher level API:

https://github.com/eclipse-ecal/ecal/blob/master/samples/python/measurement_read/measurement_read.py

Would be interesting to know what you are missing or would like to add.

**Comment by chengguizi** [comment]
Thanks @rex-schilasky , the last one, python high level API of measurement looks very interesting. I have also found its documentation here

https://eclipse-ecal.github.io/ecal/_api_python/ecal.measurement.measurement.html

However from its description, only extract (deserialize) is currently supported. May i know if there is ability in Python to write .hdf5 measurements? Similar to rosbag.write(topic, ts, obj)

**Comment by chengguizi** [comment]
Hi eCAL team,

My question still stands, while the `ecal.measurement.measurement` package in Python looks really convenient in reading / desealising the `protobuf content of an eCAL measurement`; is there a plan to add the functionality in reverse: writeing / serialising timestamp & data payload into ecal protobuf and save a proper hdf5 files?

Probably need to extend the functionality of this. As an outsider, I have little idea how to proceed.
https://github.com/eclipse-ecal/ecal/blob/master/lang/python/ecalhdf5/ecal/measurement/measurement.py
